### ISSUE TITLE
`useless_let_if_seq` ignores pattern if any expr is between variable declaration and conditional assignment.

### ISSUE THREAD

**Issue by obellish** [rust-lang/rust-clippy#11655]
### Summary

Lint is ignored if any expression (pure or not) is between a variable declaration and the conditional assignment.

### Lint Name

useless_let_if_seq

### Reproducer

I tried this code:

```rust
let foo;
let _ = bar();

if bar() {
    foo = 42;
} else {
    foo = 0;
}

let mut baz = None;

if bar() {
    baz = Some(42);
}
```

I expected to see this happen:
```
warning: `if _ { .. } else { .. }` is an expression
  --> src/lib.rs:5:5
   |
5  | /     let foo;
6  | |     let _ = bar();
7  | |
8  | |     if bar() {
...  |
11 | |         foo = 0;
12 | |     }
   | |_____^ help: it is more idiomatic to write: `let foo = if bar() { 42 } else { 0 };`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_let_if_seq
note: the lint level is defined here
  --> src/lib.rs:1:9
   |
1  | #![warn(clippy::useless_let_if_seq)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: `if _ { .. } else { .. }` is an expression
  --> src/lib.rs:14:5
   |
14 | /     let mut baz = None;
15 | |
16 | |     if bar() {
17 | |         baz = Some(42);
18 | |     }
   | |_____^ help: it is more idiomatic to write: `let <mut> baz = if bar() { Some(42) } else { None };`
   |
   = note: you might not need `mut` at all
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_let_if_seq

warning: `playground` (lib) generated 2 warnings
```

Instead, this happened:
Only the second warning appeared, due to the statement between the declaration and assignment.

### Version

```text
rustc 1.75.0-nightly (bf9a1c8a1 2023-10-08)
binary: rustc
commit-hash: bf9a1c8a193fc373897196321215794c8bebbeec
commit-date: 2023-10-08
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.2
```

**Comment by J-ZhengLi** [comment]
Failed to reproduce on [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1a454d970813b02cd075f0988b430b6f), perhaps this already got fixed?
